src/server/pfs/fuse/server.go

```diff
@@ -0,0 +1,850 @@
+			mode, ok := vs["mode"]
+			if !ok {
+				http.Error(w, "no mode", http.StatusBadRequest)
+				return
+			}
+			name, ok := vs["name"]
+			if !ok {
+				http.Error(w, "no name", http.StatusBadRequest)
+				return
+			}
+			k, ok := vs["key"]
+			if !ok {
+				http.Error(w, "no key", http.StatusBadRequest)
+				return
+			}
+			key, err := mountKeyFromString(k)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+			if key.Commit != "" {
+				http.Error(
+					w,
+					"don't support mounting commits yet",
+					http.StatusBadRequest,
+				)
+				return
+			}
+			// TODO: use response (serialize it to the client, it's polite to hand
+			// back the object you just modified in the API response)
+			_, err = mm.MountBranch(key, name, mode)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		})
+	router.Methods("PUT").
+		Queries("name", "{name}").
+		Path("/repos/{key:.+}/_unmount").HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		vs := mux.Vars(req)
+		k, ok := vs["key"]
+		if !ok {
+			http.Error(w, "no key", http.StatusBadRequest)
+			return
+		}
+		name, ok := vs["name"]
+		if !ok {
+			http.Error(w, "no name", http.StatusBadRequest)
+			return
+		}
+		key, err := mountKeyFromString(k)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		_, err = mm.UnmountBranch(key, name)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	})
+	// TODO: implement _commit
+
+	// TODO: switch http server for gRPC server and bind to a unix socket not a
+	// TCP port (just for convenient manual testing with curl for now...)
+	// TODO: make port and bind ip parameterizable
+	srv := &http.Server{Addr: ":9002", Handler: router}
+
+	go func() {
+		err := mm.Start()
+		if err != nil {
+			fmt.Printf("Error running mount manager: %s\n", err)
+			os.Exit(1)
+		}
+		err = mm.Cleanup()
+		if err != nil {
+			fmt.Printf("Error cleaning up mount manager: %s\n", err)
+			os.Exit(1)
+		}
+		srv.Shutdown(context.Background())
+	}()
+
+	return srv.ListenAndServe()
+}
+
+type MountState struct {
+	Name       string   `json:"name"`       // where to mount it. written by client
+	MountKey   MountKey `json:"mount_key"`  // what to mount. written by client
+	Mode       string   `json:"mode"`       // "ro", "rw", or "" if unknown/unspecified. written by client
+	State      string   `json:"state"`      // "unmounted", "mounting", "mounted", "pushing", "unmounted", "error". written by fsm
+	Status     string   `json:"status"`     // human readable string with additional info wrt State, e.g. an error message for the error state. written by fsm
+	Mountpoint string   `json:"mountpoint"` // where on the filesystem it's mounted. written by fsm. can also be derived from {MountDir}/{Name}
+}
+
+type MountStateMachine struct {
+	MountState
+	manager   *MountManager
+	mu        sync.Mutex
```

Comment:
I'm still assembling my mental image of how this all works, but I don't totally understand the purpose of this mutex. My understanding is that `MaybeStartFsm` uses `manager.mu` and `manager.States` to ensure that there's one goro (and one instance of `Run()`) running for a given `name` at a time. Does anything besides the state functions write to `MountStateMachine`? Could two state functions ever write to the same `MountStateMachine` concurrently, despite `manager.States` guarding `Run()`? Or is the intent to prevent reading partial writes?

Reply:
There's a mutex for the global MountManager and a mutex for each MountStateMachine. The MountManager mutex is to guard access to the `mm.States` map (since go maps can't be accessed concurrently).

For the per mount MountStateMachine mutex:

Different parts of `MountStateMachine.MountState` are written by different parts of the system, as documented in the comments in the `MountState` struct, e.g.
```
	Mode       string   `json:"mode"`       // "ro", "rw", or "" if unknown/unspecified. written by client
	State      string   `json:"state"`      // "unmounted", "mounting", "mounted", "pushing", "unmounted", "error". written by fsm
```
(where fsm means finite state machine)

I find it's always useful to document which agent in the system is responsible for writing to a given variable, and generally not having two agents both try and write to the same.

So, there might be different agents writing to different parts of the MountState from different goroutines, so holding a mutex on writing to those values seems wise to avoid overlapping executions of code that wants to modify the state. We could probably optimize this with an RWLock but I don't think it's performance critical code: we always hold the locks for a few milliseconds at most probably.

Reply:
Ah, I didn't realize `Mode` could change after startup, I'll try to understand that part a bit better

Reply:
It doesn't right now but it could, e.g a "remount"